src/Text/Pandoc/Writers/OpenDocument.hs

```diff
@@ -635,6 +655,48 @@ inlineToOpenDocument o ils
         addNote nn
         return nn
 
+--
+mkLink :: WriterOptions -> [Text] -> [Text] -> [Text] -> Text -> Text -> Doc Text -> Doc Text
+mkLink o hrefs trefs irefs s t d =
+  let isHeaderRef ident = elem ident hrefs
+      isTableRef ident = elem ident trefs
+      isImageRef ident = elem ident irefs
+      d' = inSpanTags "Definition" d
+      headerReference ident = if isEnabled Ext_number_prefix_references o
+                              -- The number prefix on header references is empty
+                              -- as generated by pandoc - indices need to be
+                              -- refreshed for the number to be generated.
+                              then inTags False "text:bookmark-ref"
+                                          [ ("text:reference-format", "number" ),
+                                            ("text:ref-name", ident) ] mempty
+                                      <>
+                                      selfClosingTag "text:s" []
+                                      <>
+                                      inTags False "text:bookmark-ref"
```

Comment:
There is a lot of code duplication in here. This would probably also benefit from a new data representation mentioned above, as it should make it easier to fold the three`*Reference` functions into one.

Reply:
Refactored this area - no code duplication now I think.